:sparkles: Add node deletion to reboot remediation

slintes opened the pull request:
**What this PR does / why we need it**:

This PR adds node deletion to the reboot remediation strategy, because it ensures that unhealthy nodes are fenced, and workloads will be rescheduled to other nodes.
This was proposed on Slack, and it was agreed to add this to the existing reboot remediation strategy, instead of adding a new strategy.

Unfortunately this increases complexity of remediation, because instead of having 1 step only (set reboot annotation), we now have several steps, plus waiting for certain cluster conditions:

1. request power off and wait until done
2. Backup node annotations and labels
3. Delete node and wait until gone
4. Power on and wait until done
5. Restore node annottaions and labels

This is WIP, tests need to be updated / extended. Initial review would be welcome though, to see if this is going into an approvable direction, or something different is expected.

beekhof commented on baremetal/metal3remediation_manager.go:
```diff
@@ -38,15 +39,20 @@ import (
 )
 
 const (
-	rebootAnnotation = "reboot.metal3.io"
```
I'm not sure you can change this... its specified by the metal3 API

slintes replied:
I understand when I change it in the way I did it will do a power off instead o a reboot?
```
The basic annotation form (reboot.metal3.io) triggers the controller to power cycle the Host. This form has set-and-forget semantics and the controller removes the annotation once it restores power to the Host.

An advanced form (reboot.metal3.io/{key}) instructs the controller hold the Host in a PoweredOff state so that the caller can perform any required actions while the node is in a known safe state.
```

https://github.com/metal3-io/metal3-docs/blob/c5608f6540e2f542da22b43e051c6c176ea08ef9/design/baremetal-operator/reboot-interface.md#proposal

macaptain commented:
Hey there.

A couple of thoughts on the PR:
* We should implement this as an alternative remediation strategy. I know there was a message on Slack suggesting that modifying the existing strategy would be OK (https://kubernetes.slack.com/archives/CHD49TLE7/p1634217997210700?thread_ts=1634199536.209500&cid=CHD49TLE7), but I advise we extend the behavior as a new strategy instead of modifying in place. Firstly, it will minimize any disruption as we trial the new remediation strategy if we can continue to use the reboot strategy while any issues are ironed out. Secondly, it will let us test both implementations side-by-side. Thirdly, it opens the possibility of new ways of extending the behavior in future.
* It would be great to have a proposal written up. The proposal detailing the existing strategy is here: https://github.com/metal3-io/metal3-docs/blob/master/design/cluster-api-provider-metal3/capm3-remediation-controller-proposal.md. I found out the background on this PR from the comment here: https://github.com/metal3-io/metal3-docs/pull/118#issuecomment-655326761. It would be good to have this raised for discussion once more.

beekhof commented:
@macaptain We originally proposed much the same thing, but respected the community's decision to take a different path.  Since you are the one wanting to re-litigate the direction, perhaps you can be the one to write up an alternative proposal and convince the community.

/cc @zaneb and @jan-est

macaptain commented:
> @macaptain We originally proposed much the same thing, but respected the community's decision to take a different path. Since you are the one wanting to re-litigate the direction, perhaps you can be the one to write up an alternative proposal and convince the community.
> 
> /cc @zaneb and @jan-est

I'm not sure I've been understood here? I don't want to "re-litigate" any directions. I see there are advantages to the remediation strategy proposed in this PR. I'd just like more of a write-up on these advantages so that I can judge if I'd like to use this remediation strategy myself, and to have the option of continuing to use the existing remediation behavior of rebooting if I choose.

Thanks.

zaneb commented:
> We should implement this as an alternative remediation strategy. I know there was a message on Slack suggesting that modifying the existing strategy would be OK (https://kubernetes.slack.com/archives/CHD49TLE7/p1634217997210700?thread_ts=1634199536.209500&cid=CHD49TLE7), but I advise we extend the behavior as a new strategy instead of modifying in place. Firstly, it will minimize any disruption as we trial the new remediation strategy if we can continue to use the reboot strategy while any issues are ironed out. Secondly, it will let us test both implementations side-by-side. Thirdly, it opens the possibility of new ways of extending the behavior in future.

I respectfully disagree here. In my opinion the lack of Node removal is a bug, and forcing users to choose between the buggy strategy and the non-buggy strategy is not doing anybody any favours. As far as disruption goes, CAPM3 has releases and nobody is obliged to chase `main`.

However, your request to have this documented so that everyone can understand why the current behaviour is buggy is entirely reasonable. A design proposal to describe the change and its benefits would be very welcome.

kashifest commented:
> I respectfully disagree here. In my opinion the lack of Node removal is a bug, and forcing users to choose between the buggy strategy and the non-buggy strategy is not doing anybody any favours. As far as disruption goes, CAPM3 has releases and nobody is obliged to chase `main`.

Isn't it so that Node removal workflow is part of the CAPI Machine health check workflow and it will remove the machine which in turn will remove the node as well which is why node deletion was not included here in the first place? I might be wrong but but it will be worth checking before we add that feature here. And I am also not sure how CAPI would behave in case we delete the node here in CAPM3 instead of CAPI doing it. It might just the delete the machine actually but as I said it is worth looking into the CAPI machine health check workflow. 
  
> However, your request to have this documented so that everyone can understand why the current behaviour is buggy is entirely reasonable. A design proposal to describe the change and its benefits would be very welcome.

I am +1 on this since this is changing the existing behavior of the controller and that's the norm we follow in meta3 generally. If you think @beekhof you wont manage time to do propose changes to the initial proposal, we can arrange a joint session and someone else can also take of that part.

zaneb commented on baremetal/metal3remediation_manager.go:
```diff
@@ -38,15 +39,20 @@ import (
 
 const (
-	rebootAnnotation = "reboot.metal3.io"
+	powerOffAnnotation              = "reboot.metal3.io/metal3-remediation-requested-power-off"
```
The key here shouldn't be a constant, but rather something unique like the UID of the remediation object.

zaneb commented:
> Isn't it so that Node removal workflow is part of the CAPI Machine health check workflow and it will remove the machine which in turn will remove the node as well which is why node deletion was not included here in the first place? I might be wrong but it will be worth checking before we add that feature here. And I am also not sure how CAPI would behave in case we delete the node here in CAPM3 instead of CAPI doing it. It might just the delete the machine actually but as I said it is worth looking into the CAPI machine health check workflow.

The standard remediation workflow in CAPI is indeed to delete the Machine, which has the effect of deleting the Node.

Remediation controllers are an alternative to the standard remediation workflow. The purpose of the 'reboot' remediation strategy is to allow us to **not** delete the Machine (which would mean reprovisioning the Host) and instead just reboot the Host. Because the Machine itself remains, this means that the CAPI will not delete the Node for us, which is an essential part of remediation, and thus the remediation controller (_all_ remediation controllers) _must_ do it. That's why it's a bug that we currently don't.

macaptain commented:
> I respectfully disagree here. In my opinion the lack of Node removal is a bug

Hey @zaneb. Is there an issue describing the defect? I couldn't find an issue mentioning the remediation controller in CAPM3 repo. You will hopefully forgive me for thinking that a draft PR named ":sparkles: Add ..." with no linked issue is for a new feature rather than a bugfix. (Sadly, I do need these details to be made clear for me, I'm not a mind reader - although if I were, it sure would help me write other people's change proposals @beekhof hehe)

From the perspective that this is a new feature PR, hopefully you'll also understand why I thought an open-closed-principle-be-damned approach, with +300, -50 into the heart of a controller from an [implemented proposal](https://github.com/metal3-io/metal3-docs/blob/11cc01510e3936fbefcab818419f0b4a249287de/design/cluster-api-provider-metal3/capm3-remediation-controller-proposal.md?plain=1#L14) may be best behind a configuration setting. 

There may still be advantages to putting a significant bugfix behind a feature flag. It means we can merge to `main` and then follow up on getting any e2e tests right while iterating on the fix, so I'll put that idea forward for consideration again. The idea is we can deprecate the old behaviour once we're sure that the new way of doing things is better in every way.

> forcing users to choose between the buggy strategy and the non-buggy strategy is not doing anybody any favours. As far as disruption goes, CAPM3 has releases and nobody is obliged to chase `main`.

If there's something that surely does nobody any favours it's users stuck on old releases because `main` doesn't do what they want anymore. Let's not end up there!

You seem convinced that this change is an unalloyed benefit and that the previous approach specified in the remediation controller proposal should be eliminated. This is entirely possible and I would really like to understand more details. In particular, I would appreciate:

1. An issue so that I can explain to my customers that they have a defect in their environment in the remediation controller.
2. A change proposal so that I can explain to my customers all the benefits of the new way of doing things.

Cheers!

slintes commented:
I don't care that much if we call it a bugfix or feature, but I tend to bugfix, because of Nir's comment on the old PR and Zane's comments here. The thing I still need to know however if this should be a new strategy or not, no matter how we call it in the end...

> The proposal detailing the existing strategy...

Actually, I don't see any details there about how remediation will happen, it jut says "RC uses BMO APIs to get hosts back into a healthy or manageable state" 🤷🏼 But I'm happy to add the details. 

>  I couldn't find an issue 

Or I create an issue. Whatever you prefer... or both.

> a draft PR named ":sparkles: Add ..."

Happy to fix the title and make this a bugfix. I just can't find the docs explaining the emojis anymore... where is it? 🤔 I expected it to be in https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/.github/PULL_REQUEST_TEMPLATE.md...?

>  with +300, -50 into the heart of a controller

Line count, seriously? I'm just putting node deletion between the power off and power on of the reboot. Sorry that this isn't possible with fewer lines of code 🤷🏼‍♂️ And I explained why in the decription.

>  follow up on getting any e2e tests right 

Tests will be part of this PR, it's WIP.

slintes commented:
https://github.com/metal3-io/cluster-api-provider-metal3/issues/392

macaptain commented:
> Line count, seriously? I'm just putting node deletion between the power off and power on of the reboot. Sorry that this isn't possible with fewer lines of code 🤷🏼‍♂️ And I explained why in the decription.

Sorry this comment came across like this. I really do appreciate the improvements that this change will bring to remediation. It can only be a good thing to have node deletion at least as an option. Also having this as the only remediation behavior and replacing the old may be far better, I just need to check the reasons for that.

> #392

Thanks for raising the issue. Having the issue up as a place for discussion is appreciated. @zaneb and I had a discussion in the community meeting, and all's well. We agreed having more documentation around remediation would be a good thing since remediation has some history in the project which I myself being new to this area am also unaware of. We spoke about a 'change proposal', which I think may be a good thing as well? Since consensus based on what someone said on Slack means we don't have all the technical background documented.

> Tests will be part of this PR, it's WIP.

Great, thanks!! I was also planning on writing some unit tests for the old behavior since I noticed there aren't any. Hopefully will have a PR up before too long and we can compare notes. 

There's also an open PR in the metal3-dev-env repo where some remediation behavior was being tested: https://github.com/metal3-io/metal3-dev-env/pull/828. I was going to take a closer look at that PR and see if it's something we want to merge and add any useful tests there.

Also, welcome to metal3! Sorry - the discussions on PRs are not usually this robust, this PR has been interesting :laughing:

zaneb commented:
> sure... where and how do I file a change proposal? thinking

PR to the [metal3-docs](https://github.com/metal3-io/metal3-docs/) repo. The original design was https://github.com/metal3-io/metal3-docs/blob/master/design/cluster-api-provider-metal3/capm3-remediation-controller-proposal.md - we could edit that one but since it is marked as 'implemented' already it might be better to create a new proposal.

macaptain commented:
> just to be sure: does this mean you are also fine with adding node deletion to the existing remdiation strategy?

I overestimated the maturity of the old code and implementation. After reading the comments on the issue, I changed my mind and I think it would be better to change the old strategy in place and fix the issue directly.

Just so I could get my head around what the remediation controller does at the moment, I raised a draft PR to add a few unit tests covering the remediation steps: https://github.com/metal3-io/cluster-api-provider-metal3/pull/404. Let me know what you think!

> Btw, I'm still busy with the e2e tests here, having a hard time to actually run them... I might reach out on Slack for help.

Sure! In #cluster-api-baremetal Kubernetes Slack there's several people who know a good deal about the e2e tests and would be happy to help.

> it might be better to create a new proposal.

As @slintes said, the old proposal actually doesn't include details on how the remediation controller actually remediates the machine. The proposal has been implemented nonetheless, so I think a new proposal is easiest. Not all the template questions may be relevant. For me and my ignorance, covering the steps in documentation of what remediation should actually do in CAPM3 would be very beneficial. Thank you!

smoshiur1237 commented:
Hi @slintes, This PR has been staying as a draft from end of november 2021.  Do you any update on this PR and any plan to add e2e test?

slintes commented:
Hi @smoshiur1237, we are currently finalizing our plannings for the next weeks and months. I can probably continue with this soon. If not, I will let you know. Thanks.

slintes commented:
rebased, looking into tests again :)

slintes commented:
/test all

furkatgofurov7 commented:
/ok-to-test

slintes commented:
Fixed generate, lint and vet issues, updated remediation unit test.
Will extend manager unit test for new methods.
Also still looking into e2e tests, at least got a metal3-dev-env cluster running already :)

slintes commented:
/test all

slintes commented:
Added missing unit tests, and a new e2e test, and fixed errors I found during writing tests 😎 
Ready for review 🙂 

/test all

slintes commented:
/test all

slintes commented:
/test-v1b1-centos-integration
/test-v1b1-ubuntu-integration

slintes commented:
/test all

zaneb commented on controllers/metal3remediation_controller.go:
```diff
@@ -222,6 +220,250 @@ func (r *Metal3RemediationReconciler) reconcileDelete(ctx context.Context,
+		if set := remediationMgr.SetFinalizer(); set {
+			now := metav1.Now()
+			remediationMgr.SetLastRemediationTime(&now)
+			remediationMgr.IncreaseRetryCount()
```
The finalizer (which is in the object metadata) and Status data (in a subresource) are written to the API by [two separate API calls](https://github.com/kubernetes-sigs/cluster-api/blob/main/util/patch/patch.go#L142-L143).
Therefore it's not safe to rely on the state from the first one to ensure the second one only happens once. If the second write fails then it will never be repeated.

I think you want to update the count/time atomically with the change into the Starting phase.

slintes replied:
good catch, thanks!

slintes commented on api/v1beta1/metal3remediation_types.go:
```diff
@@ -32,12 +32,21 @@ const (
-	// PhaseRunning represents the running state during remediation.
-	PhaseRunning = "Running"
+	// PhaseStarting represents the state when remediation started.
+	PhaseStarting = "Starting"
```
tbh I didn't consider that this might be an API change, thanks for the heads up

slintes commented on api/v1beta1/metal3remediation_types.go:
```diff
@@ -32,12 +32,21 @@ const (
+	PhasePoweredOff = "PoweredOff"
+
+	// PhasePoweredOn represents the state when the host is powered on again.
+	PhasePoweredOn = "PoweredOn"
```
I managed to revert to the old phases 👍🏼

slintes commented on controllers/metal3remediation_controller.go:
```diff
@@ -141,76 +149,66 @@ func (r *Metal3RemediationReconciler) reconcileNormal(ctx context.Context,
-					remediationMgr.SetLastRemediationTime(&now)
-					remediationMgr.IncreaseRetryCount()
-				}
+			return r.remediateRebootStrategy(ctx, remediationMgr)
```
I had 2 commits in an earlier PR version, but squashed them after I had to big changes in both `reconcileNormal()` and `remediateRebootStrategy()`... it wasn't helpful anymore IMHO.
I will see if I can get back to 2 commits after reverting to the old phases...

slintes commented:
@zaneb thanks for review! I fixed issues locally, but fail to run the e2e test on my server now (cluster issues) :/
Will push changes when I get the e2e test running again.
And I just see I need to rebase.

slintes commented:
PR updated, I got my dev cluster working again, somehow I had a broken ironic image in the local registry.

zaneb commented on baremetal/metal3remediation_manager.go:
```diff
@@ -289,3 +356,130 @@ func (r *RemediationManager) GetCapiMachine(ctx context.Context) (*clusterv1.Mac
+
+// UpdateNode updates the given node.
+func (r *RemediationManager) UpdateNode(ctx context.Context, node *corev1.Node) error {
+	clusterClient, err := r.getClusterClient(ctx)
```
It might be more efficient to pass the client in, so that e.g. we don't have to create it twice to get and then update the Node.

slintes replied:
done

zaneb commented on controllers/metal3remediation_controller.go:
```diff
@@ -43,6 +46,7 @@ type Metal3RemediationReconciler struct {
 // +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3remediations,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3remediations/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines;machines/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;create;update;patch;delete
```
create?

slintes replied:
oh, good catch

zaneb commented on controllers/metal3remediation_controller.go:
```diff
@@ -141,76 +149,96 @@ func (r *Metal3RemediationReconciler) reconcileNormal(ctx context.Context,
 			remediationMgr.SetRemediationPhase(infrav1.PhaseRunning)
+			now := metav1.Now()
+			remediationMgr.SetLastRemediationTime(&now)
+			remediationMgr.IncreaseRetryCount()
```
This looks a little odd, given that we haven't performed a remediation yet but we're setting a "last remediation time" and a retry count of 1. I guess we're reusing existing fields to have slightly different meanings. The latter at least seems unnecessary.

What is the impact on upgrades from previous versions of the controller (i.e. when there are remediations in flight)?

slintes replied:
removed `IncreaseRetryCount`

zaneb commented on controllers/metal3remediation_controller.go:
```diff
@@ -141,76 +149,96 @@ func (r *Metal3RemediationReconciler) reconcileNormal(ctx context.Context,
+		}
+
+		// try to get node
+		node, err := remediationMgr.GetNode(ctx)
```
It looks to me like `getClusterClient()` is potentially getting a client for another cluster.
Having this here means that e.g. if the Node is in another cluster and we don't have RBAC access to it (wasn't required before) then we won't be able to do _anything_ to remediate. Is it possible to avoid this error up to the point where we actually need the Node?

slintes replied:
done

slintes replied:
do I understand correctly, it can be that capm3 is updated, but updated RBAC roles won't apply for existing clusters?

slintes replied:
I'm handling the error now by skipping node releated actions

zaneb commented on controllers/metal3remediation_controller.go:
```diff
@@ -141,76 +149,96 @@ func (r *Metal3RemediationReconciler) reconcileNormal(ctx context.Context,
-					r.Log.Error(err, "error setting cluster api conditions")
-					return ctrl.Result{}, errors.Wrapf(err, "error setting cluster api conditions")
-				}
+					// Remediation failed to set unhealthy annotation on BMH
```
```suggestion
					// Remediation failed, so set unhealthy annotation on BMH
```

zaneb commented on controllers/metal3remediation_controller.go:
```diff
@@ -222,6 +250,203 @@ func (r *Metal3RemediationReconciler) reconcileDelete(ctx context.Context,
+// Return a Result and optionally an error when reconcile should return.
+func (r *Metal3RemediationReconciler) remediateRebootStrategy(ctx context.Context,
+	remediationMgr baremetal.RemediationManagerInterface, node *corev1.Node) (ctrl.Result, error) {
+	switch remediationMgr.GetRemediationPhase() {
```
This is only called from the Running phase, what's the point of checking it at runtime?

slintes replied:
left over from old version, removed, thx

zaneb commented on controllers/metal3remediation_controller.go:
```diff
@@ -222,6 +250,203 @@ func (r *Metal3RemediationReconciler) reconcileDelete(ctx context.Context,
+			return ctrl.Result{}, errors.Wrap(err, "error getting poweroff annotation status")
+		} else if ok {
+			r.Log.Info("Powering on the host")
+			err := remediationMgr.RemovePowerOffAnnotation(ctx)
```
You shouldn't do this until after the phase has been changed to Waiting, because if the write to the Status fails then we'll end up re-adding another reboot annotation.

slintes replied:
done

zaneb reviewed (commented):
If you're not going to split the first commit then it should at least not advertise itself as a 'refactor'.

slintes commented:
> If you're not going to split the first commit then it should at least not advertise itself as a 'refactor'.

rephrased

> What is the impact on upgrades from previous versions of the controller (i.e. when there are remediations in flight)?

Good question. I expect CRs in running phase would be remediated completly, because it doesn't find the power off annotation. For CRs in waiting phase it would just wait and check timeout / retry count, and potentially start a new remediation. So IMHO we are fine. Except for the potential node RBAC issue, IIUC?

slintes commented:
Handlng missing RBAC for nodes now, it will skip node backup / deletion / restore

slintes commented:
Rebased.
I added the new Metal3Remediation e2e test to the `[remediation]` labeled test. Let me know if you want it to have it its own label.

Rozzii commented:
/cc @smoshiur1237

zaneb commented:
/test-v1b1-ubuntu-integration

smoshiur1237 commented:
Thanks for the fix and it looks good to me. But I prefer to have more review.
/cc @mboukhalfa 
/cc @furkatgofurov7 
/cc @fmuyassarov 
/test-v1b1-ubuntu-e2e

Xenwar commented:
I have done a test and noticed the following.

1. k8s node is not deleted
2. workload relocation is not relocated, but re-created after reboot
3. labels are restored after reboot
4. annotations are restored after reboot

I have used the following command to test it, could there be something else to add before testing it. 
`kubectl annotate bmh node-0 -n metal3 reboot.metal3.io=
`

slintes commented:
@Xenwar thanks for having a look.

> kubectl annotate bmh node-0 -n metal3 reboot.metal3.io=

I don't think this is a good test. I modified the controller which reconciles remediation CRs. So you need to create such a CR. See example in the e2e test: https://github.com/metal3-io/cluster-api-provider-metal3/pull/367/files#diff-51dc21f2a53490d952b41eb5bfd84154f85ba1a691c5a7173ea4548d3aea10d5R42

furkatgofurov7 commented:
@slintes sorry for the delay, thanks for the patch, I will try to spend some time this week going through it. However, I quickly checked it and seems after these changes code coverage for controllers is decreasing. Generally we try to keep it above 80% overall for all packages as per https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/docs/testing.md

lentzi90 commented on docs/e2e-test.md:
```diff
@@ -51,8 +51,8 @@ make clean -C /opt/metal3-dev-env/metal3-dev-env/ && sudo rm -rf /opt/metal3-dev
 
 The e2e tests currently include three different sets:
 
-1. pivoting_based_feature_tests
-1. remediation_feature_tests
```
Why rename these? AFAIK there are multiple tests in each so it makes sense to call them `tests` rather than `test`.

slintes replied:
TBH I don't remember if I had a real issue with `_tests.go`, or if just Goland showed a mix of normal and test files in that directory because of the different suffixes, by convention all test files in Go should have the `_test.go` suffix. And most of them contain multiple tests.

lentzi90 replied:
Good point, thanks!

slintes replied:
np 🙂

lentzi90 reviewed (commented):
Found a couple of nits below. Otherwise I think this looks pretty good!
I would suggest squashing into no more than 3 commits though before merging.

kashifest commented:
@slintes can you please readdress the comments and lets take this in afterwards

slintes commented on baremetal/metal3remediation_manager.go:
```diff
@@ -141,11 +161,58 @@ func (r *RemediationManager) SetRebootAnnotation(ctx context.Context) error {
+		return false, err
+	}
+	if host == nil {
+		return false, errors.New("Unable to remove PowerOff Annotation, Host not found")
```
fixed, thx

slintes commented on baremetal/metal3remediation_manager.go:
```diff
@@ -141,11 +161,58 @@ func (r *RemediationManager) SetRebootAnnotation(ctx context.Context) error {
+		return false, err
+	}
+	if host == nil {
+		return false, errors.New("Unable to remove PowerOff Annotation, Host not found")
```
fixed, thx

kashifest commented:
> Rebased and addressed comments. Sorry for the delay, too busy with our next release...

No problem and thanks for the update. I will check it tomorrow hopefully, and dont worry about the failed broken link test. I am fixing it in another PR and I will rerun the tests once that is merged

slintes commented:
>  dont worry about the failed broken link test

Too late, already fixed in last commit 😉 Should I revert?

kashifest commented:
> > dont worry about the failed broken link test
> 
> Too late, already fixed in last commit wink Should I revert?

I think the fix landed already. You might need to rebase. Commit list is a bit heavy. I will leave that to @slintes to decide if we can squash a few of those. @furkatgofurov7 your verdict would be final regarding the test coverage. Otherwise PR looks good to me. Thanks @slintes for taking care of this.

/approve

furkatgofurov7 commented on test/e2e/metal3remediation_test.go:
```diff
@@ -0,0 +1,98 @@
+	waitForNodeStatus(ctx, targetClient, client.ObjectKey{Name: workerNodeName}, corev1.ConditionTrue, specName)
+
+	By("Deleting remediation")
+	Expect(bootstrapClient.Delete(ctx, m3Remediation)).To(Succeed(), "should delete Metal3Remediation")
```
```suggestion
	Expect(bootstrapClient.Delete(ctx, m3Remediation)).To(Succeed(), "should delete Metal3Remediation CR")
```

furkatgofurov7 commented on test/e2e/metal3remediation_test.go:
```diff
@@ -0,0 +1,98 @@
+	By("Waiting for node ready")
+	waitForNodeStatus(ctx, targetClient, client.ObjectKey{Name: workerNodeName}, corev1.ConditionTrue, specName)
+
+	By("Deleting remediation")
```
```suggestion
	By("Deleting Metal3Remediation CR")
```

furkatgofurov7 commented on controllers/metal3remediation_controller.go:
```diff
@@ -141,76 +150,138 @@ func (r *Metal3RemediationReconciler) reconcileNormal(ctx context.Context,
+					remediationMgr.RemoveNodeBackupAnnotations()
+					remediationMgr.UnsetFinalizer()
+
+					r.Log.Info("Node restored, remediation done, CR should be deleted soonish")
```
maybe just: 'soon' :)

slintes replied:
ok 😉

furkatgofurov7 reviewed (commented):
Hi @slintes, thanks for the PR, this is great work! I went over the long discussion here and seems we are missing some points here:

1. Have we decided already what are the plans with changing existing proposal documentation which lives in metal3-docs repo, are you planning to file a new proposal or just change the existing one with the implementation details from this PR?
2. We do have a similar ansible based remediation tests in metal3-dev-env repo that was mentioned on this thread, and AFAIK that needs to be accordingly updated with the extra steps? That is not a hard requirement tbh, since we are on the way of replacing those tests with e2e tests on this repo, so I will leave this discussion open to others. /cc @kashifest @smoshiur1237 
3. The controllers package test coverage % is slightly dropping after these changes, it would be great to see some improvement on that as well. 

Minor improvement comments below:

kashifest commented:
/test-e2e-ubuntu-parallel

kashifest commented:
/test-v1b1-e2e-ubuntu-parallel

slintes commented:
Thx for your review @furkatgofurov7 

> Hi @slintes, thanks for the PR, this is great work! I went over the long discussion here and seems we are missing some points here:
> 
> 1. Have we decided already what are the plans with changing existing proposal documentation which lives in metal3-docs repo, are you planning to file a new proposal or just change the existing one with the implementation details from this PR?

I mentioned this earlier:
> Actually, I don't see any details there about how remediation will happen, it jut says "RC uses BMO APIs to get hosts back into a healthy or manageable state"

If you like me to add implementation details (since there none yet I guess a new proposal does not make sense), I can do that as follow up.

> 2. We do have a similar ansible based remediation tests in metal3-dev-env repo that was mentioned on this thread, and AFAIK that needs to be accordingly updated with the extra steps? That is not a hard requirement tbh, since we are on the way of replacing those tests with e2e tests on this repo, so I will leave this discussion open to others. /cc @kashifest @smoshiur1237

Tbh I forgot about this, sorry. But when it's replaced anyway, I'd prefer to not spend time on it (I'm not that familiar with Ansible, so I'm afraid it would take more time than expected)
 
> 3. The controllers package test coverage % is slightly dropping after these changes, it would be great to see some improvement on that as well.

I think I added tests for all new functions, but I will have a look what I missed...

> 
> Minor improvement comments below:

checking now 🙂

lentzi90 commented on test/e2e/metal3remediation_test.go:
```diff
@@ -0,0 +1,98 @@
+	}
+	Expect(bootstrapClient.Create(ctx, m3Remediation)).To(Succeed(), "should create Metal3Remediation")
+
+	By("Waiting for power off")
```
I don't think we should say "VM" here. Maybe "machine"?

slintes commented on baremetal/mocks/zz_generated.metal3remediation_manager.go:
```diff
@@ -30,7 +30,9 @@ import (
 	v1beta1 "github.com/metal3-io/cluster-api-provider-metal3/api/v1beta1"
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/api/core/v1"
+	v10 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v11 "k8s.io/client-go/kubernetes/typed/core/v1"
```
me neither, but I also never really cated 😉

slintes commented:
feedback addressed, will check test coverage and squash commits hopefully later today

slintes commented:
/test test-ubuntu-integration-main

kashifest commented:
> > 2. We do have a similar ansible based remediation tests in metal3-dev-env repo that was mentioned on this thread, and AFAIK that needs to be accordingly updated with the extra steps? That is not a hard requirement tbh, since we are on the way of replacing those tests with e2e tests on this repo, so I will leave this discussion open to others. /cc @kashifest @smoshiur1237
> 
> Tbh I forgot about this, sorry. But when it's replaced anyway, I'd prefer to not spend time on it (I'm not that familiar with Ansible, so I'm afraid it would take more time than expected)


I think its fine. Creating an issue would be enough so that we dont forget about it.

slintes commented:
Improved test coverage of remediation controller, almost only error handling is uncovered now.
Also squashed some commits.

slintes commented:
/test test-ubuntu-integration-main

furkatgofurov7 commented:
/test-ubuntu-e2e-parallel-main

furkatgofurov7 commented:
/test-centos-e2e-parallel-main

kashifest commented:
/test-ubuntu-integration-main

furkatgofurov7 commented:
/retest

furkatgofurov7 commented:
/retest

adilGhaffarDev commented:
/test shellcheck

furkatgofurov7 commented:
/test govet

furkatgofurov7 commented:
/test all

kashifest commented:
/test-ubuntu-integration-main

furkatgofurov7 commented:
@slintes can you please rebase on top of the main branch, we are seeing some problems running the tests, thanks!

mboukhalfa commented:
/test-ubuntu-integration-main

mboukhalfa commented:
/ok-to-test

mboukhalfa commented:
/test-ubuntu-integration-main

kashifest commented:
/test-ubuntu-integration-main

Rozzii commented:
/test-ubuntu-integration-main

furkatgofurov7 commented:
/test all
/test-ubuntu-integration-main

furkatgofurov7 commented:
/test-centos-e2e-parallel-main
/test-centos-e2e-main

furkatgofurov7 commented:
@slintes hi! triggering the CI in this specific PR does not work for some unknown reason we are trying to find out but does work in any other PR in this repo. What I would suggest is, for now, can you please open a new PR with same changes on top of the main branch (sorry for extra work :slightly_smiling_face: ), because we had seen similar issue in the past and simple solution was creating a new one (not 100% guarantee it will work, but we can give it a try). Meanwhile I will continue investigating the root cause of the issue. Thanks.

slintes commented:
https://github.com/metal3-io/cluster-api-provider-metal3/pull/668

furkatgofurov7 commented:
> #668

Thanks, seems to work now :smile:

furkatgofurov7 commented:
/close
